### DISCUSSION TITLE
Module linker

### DISCUSSION THREAD

**Discussion by klntsky** [mlabs-haskell/purus#4] [General]
Let's discuss how to handle modules.

PureScript compiles code on per-module basis, because it has to maintain per-module FFI interfaces.

However, in our context we are always interested in just one function - a validator script.

We need to find a way to tie modules together.

Let's start with an obvious approach and then try to improve on it.

I propose embedding modules into CoreFn. Conceptually, a module can be seen as a function from a record of its imports to a record of its exports. Modules can't be co-recursive, so every valid module set can be transformed into a chain of let-bindings. 

The downside is that modules are not zero-cost with this approach. Can we improve it?

**Comment by gnumonik** [comment]
In light of some conversations with the IOG folks, I'd like to propose an alternative: 

  1. Instead of modifying the PureScript compiler to emit PLC directly, we modify it to emit PIR "Modules", where a PIR Module is (essentially) a list of [PIR Bindings](https://github.com/input-output-hk/plutus/blob/caedd02422a264b1c529a5ad60537401678cdba1/plutus-core/plutus-ir/src/PlutusIR/Core/Type.hs#L93-L96). (PIR "Modules" don't exist, I'm speculating about a data structure we'd have to create) 
  2. We implement our own, external, final compilation machinery (i.e. we don't modify the PS compiler to do this) which operates on PIR modules.  

I *think* that it might be possible to derive a PIR "Module" from a PS `ExternsFile` (which is a build artifact produced regardless of the codegen target). If that is in fact possible, then we don't need to _modify_ the PS compiler's compilation pipeline at all, which would be a great boon for maintainability.

**Reply by gnumonik** [reply]
Oh and also, all of their optimization tools work on PIR and I believe it is generally the case that PIR is amenable to more/better optimizations than we could do with PLC
